shell/platform/embedder/embedder_external_view_dl_op_spy.cc

```diff
@@ -0,0 +1,159 @@
+  FML_DLOG(ERROR) << "drawRect ";
+}
+void EmbedderExternalViewDlOpReceiver::drawOval(const SkRect& bounds) {}
+void EmbedderExternalViewDlOpReceiver::drawCircle(const SkPoint& center,
+                                                  SkScalar radius) {}
+void EmbedderExternalViewDlOpReceiver::drawRRect(const SkRRect& rrect) {}
+void EmbedderExternalViewDlOpReceiver::drawDRRect(const SkRRect& outer,
+                                                  const SkRRect& inner) {}
+void EmbedderExternalViewDlOpReceiver::drawPath(const SkPath& path) {
+  FML_DLOG(ERROR) << "drawPath ";
+}
+void EmbedderExternalViewDlOpReceiver::drawArc(const SkRect& oval_bounds,
+                                               SkScalar start_degrees,
+                                               SkScalar sweep_degrees,
+                                               bool use_center) {}
+void EmbedderExternalViewDlOpReceiver::drawPoints(PointMode mode,
+                                                  uint32_t count,
+                                                  const SkPoint points[]) {}
+void EmbedderExternalViewDlOpReceiver::drawVertices(const DlVertices* vertices,
+                                                    DlBlendMode mode) {
+  FML_DLOG(ERROR) << "drawVertices ";
+}
+void EmbedderExternalViewDlOpReceiver::drawImage(const sk_sp<DlImage> image,
+                                                 const SkPoint point,
+                                                 DlImageSampling sampling,
+                                                 bool render_with_attributes) {
+  FML_DLOG(ERROR) << "drawImage ";
+}
+void EmbedderExternalViewDlOpReceiver::drawImageRect(
+    const sk_sp<DlImage> image,
+    const SkRect& src,
+    const SkRect& dst,
+    DlImageSampling sampling,
+    bool render_with_attributes,
+    SrcRectConstraint constraint) {
+  FML_DLOG(ERROR) << "drawImageRect ";
+}
+void EmbedderExternalViewDlOpReceiver::drawImageNine(
+    const sk_sp<DlImage> image,
+    const SkIRect& center,
+    const SkRect& dst,
+    DlFilterMode filter,
+    bool render_with_attributes) {}
+void EmbedderExternalViewDlOpReceiver::drawAtlas(const sk_sp<DlImage> atlas,
+                                                 const SkRSXform xform[],
+                                                 const SkRect tex[],
+                                                 const DlColor colors[],
+                                                 int count,
+                                                 DlBlendMode mode,
+                                                 DlImageSampling sampling,
+                                                 const SkRect* cull_rect,
+                                                 bool render_with_attributes) {}
+void EmbedderExternalViewDlOpReceiver::drawDisplayList(
+    const sk_sp<DisplayList> display_list,
+    SkScalar opacity) {
+  FML_DLOG(ERROR) << "drawDisplayList ";
```

Comment:
For this method, I would do something like:

```
  Something like [auto remembered_transparency_info = current_transparency_info_;]
  Something like [current_transparency_info_ = defaults;]
  display_list->Dispatch(this);
  Something like [current_transparency_info_ = remembered_transparency_info]
```

Basically, play back the DL with the following caveats:
- Your local attributes should be unaffected by the actions of the DL
- The DL should be played back assuming default attributes
- Your local attributes when everything is done should be the same state as before you recursed.

That is why I have the "save/restore" lines for whatever "current_transparency_info" might be.

Reply:
Another option would be:

```
  EmbedderExternalViewDlOpReceiver sub_receiver;
  display_list->Dispatch(sub_receiver);
  this->did_draw_ = this->did_draw_ || sub_receiver.did_draw_;
```

Reply:
And, of course, if "did_draw" is already true, there is no need to recurse in this call as it can't add any new info for your eventual reported answer...